### DISCUSSION TITLE
Add a new field to the Typegoose subdocument

### DISCUSSION THREAD

**Discussion by willmartinsmg** [typegoose/typegoose#726] [Q&A] [answered]
How do I include an extra field in the User class subdocument?

Inside the user's cars subdocument, I need a field that indicates the active or deactivated car for that user.

```javascript
class Car {
  @prop()
  public model?: string;
}

class User {
  @prop()
  public name?: string;

  @prop({ required: true })
  public age!: number; 

  @prop({ ref: () => Car })
  public cars?: Ref<Car>[]; 
}
```

Populated Car Collection:
```javascript
Car A
{
  "_id": "1"
  "model": "Ferrari"
}

Car B
{
  "_id": "2"
  "model": "Tesla"
}
```

User collection populated according to the default class:

```javascript
User
{
  "_id": "1",
  "name": "Jonh Doe",
  "age": 25,
  "cars": [
    { "_id": "1" },
    { "_id": "2" }
  ]
}
```

I need to include a "status" field in the cars array of the User collection, as shown below:

status is to identify whether the car model is active for the user

```javascript
User
{
  "_id": "1",
  "name": "Jonh Doe",
  "age": 25,
  "cars": [
    { "_id": "1", "status": false },
    { "_id": "2", "status": true }
  ]
}
```

**Comment by hasezoey** [comment, marked as answer]
You have multiple options:
- make the current reference array a subdocumentarray (see example 1)
- include the `status` property on the car itself
- have a separate class that has a unique compound index on the user-car references and has the properties you want and have such a reference field in the user (either through virtual populate or with a explicit ref array) (see example 2)

Example 1:
```ts
class Car {
  @prop()
  public model?: string;
}

class UserCarSub {
  @prop({ ref: () => Car })
  public car: Ref<Car>;

  @prop()
  public extra_properties: string;
}

class User {
  @prop()
  public name?: string;

  @prop({ required: true })
  public age!: number; 

  @prop()
  public cars?: UserCarSub[];
}
```

Example 2:
```ts
// the first example does not use virtual populate
class Car {
  @prop()
  public model?: string;
}

class UserCarProperties {
  @prop({ ref: () => Car })
  public car: Ref<Car>;

  @prop({ ref: () => User })
  public user: Ref<User>;

  @prop()
  public extra_properties: string;
}

class User {
  @prop()
  public name?: string;

  @prop({ required: true })
  public age!: number; 

  @prop({ ref: () => Car })
  public cars?: Ref<Car>[]; 

  @prop({ ref: () => UserCarProperties })
  public car_properties: Ref<UserCarProperties>[];
}

// the following uses virtual populate
class User {
  @prop()
  public name?: string;

  @prop({ required: true })
  public age!: number; 

  @prop({ ref: () => Car })
  public cars?: Ref<Car>[]; 

  @prop({ ref: () => UserCarProperties, foreignField: 'user', localField: '_id' })
  public car_properties: Ref<UserCarProperties>[]; // this property does not exist in the database, but can still populate
}
```

[Virtual Populate](https://typegoose.github.io/typegoose/docs/api/virtuals/#virtual-populate)

Alternatively you could also in Example 2 merge the cars references into the `UserCarProperties` and use that to find cars, that is fully up to how you want it

Personally i would suggest Example 2 with virtual populate (merged) if the array is of unknown length or could grow to unknown length

---

Edit: i have also found [this issue on stackoverflow](https://stackoverflow.com/questions/72737409/add-a-new-field-to-the-typegoose-subdocument), in case anyone comes across here

**Reply by willmartinsmg** [reply]
Thanks for the help and for being so clear on the solution!
